### PR TITLE
Adaptive Load exponential search step controller

### PR DESCRIPTION
**Pull request by eric846**
A StepController plugin that searches exponentially for a bad input that causes metrics to go outside thresholds, then performs a binary search between the last known good input and the bad input. It should not exceed 2x the healthy capacity of the system under test. 

The exponent of the exponential search can be adjusted from 2. 

Note that this algorithm does not take into account the magnitude of the metric score, only the sign.

This PR also adds FakeInputVariableSetter in order to use it in the unit tests for the StepController, and at the same time removes  multiple implementation inheritance for all InputVariableSetters.

### CODE REVIEW

**Comment by dubious90** [comment]
@qqustc Please review and assign back to me once complete.

**Review by oschaaf — CHANGES_REQUESTED** [review]
looks great overall; flushing out my first feedback.

File: test/adaptive_load/step_controller_test.cc [hunk 1]
```diff
@@ -0,0 +1,194 @@
...
+
+nighthawk::adaptive_load::BenchmarkResult MakeBenchmarkResultOutsideThreshold() {
+  nighthawk::adaptive_load::BenchmarkResult result;
+  nighthawk::adaptive_load::MetricEvaluation* evaluation = result.add_metric_evaluations();
+  evaluation->set_threshold_score(-1.0);
+  evaluation->set_weight(10.0);
+  return result;
+}
+
+nighthawk::adaptive_load::BenchmarkResult MakeBenchmarkResultWithinThreshold() {
```
Comment (qqustc) [inline test/adaptive_load/step_controller_test.cc:30]: It seems we can combine MakeBenchmarkResultWithinThreshold and MakeBenchmarkResultOutsideThreshold into a single MakeBenchmarkResult(bool within_threshold).

**Comment by qqustc** [comment]
> LGTM except several small nit.

Assigned back to @dubious90

File: source/adaptive_load/step_controller_impl.cc [hunk 2]
```diff
@@ -0,0 +1,161 @@
...
+}
+
+void ExponentialSearchStepController::UpdateAndRecompute(const BenchmarkResult& benchmark_result) {
+  if (benchmark_result.status().code()) {
+    doom_reason_ = "Nighthawk Service returned an error.";
+    return;
+  }
+  const double score = TotalScore(benchmark_result);
+
+  if (is_exponential_phase_) {
```
[the remark this answers is not in the document]
Comment (eric846) [reply source/adaptive_load/step_controller_impl.cc:126]: This is intriguing, see issue #470.

[the remark this answers is not in the document]
Comment (oschaaf) [reply source/adaptive_load/step_controller_impl.cc:126]: Resolving this here, thanks for filing that

File: source/adaptive_load/step_controller_impl.cc [hunk 3]
```diff
@@ -0,0 +1,161 @@
...
+    if (score > 0.0) {
+      // Have not reached the threshold yet; continue increasing the load exponentially.
+      previous_load_value_ = current_load_value_;
+      current_load_value_ *= exponential_factor_;
+    } else {
+      // We have found a value that exceeded the threshold.
+      // Prepare for the binary search phase.
+      if (std::isnan(previous_load_value_)) {
+        // Cannot continue if the initial value already exceeds metric thresholds.
+        doom_reason_ = "Outside threshold on initial input.";
```
[the remark this answers is not in the document]
Comment (eric846) [reply source/adaptive_load/step_controller_impl.cc:136]: Done

File: source/adaptive_load/step_controller_impl.cc [hunk 4]
```diff
@@ -0,0 +1,161 @@
...
+  if (!status.ok()) {
+    return status;
+  }
+  return options;
+}
+
+bool ExponentialSearchStepController::IsConverged() const {
+  // Binary search has brought successive input values within 1% of each other.
+  return doom_reason_.empty() && !is_exponential_phase_ &&
+         abs(current_load_value_ / previous_load_value_ - 1.0) < 0.01;
```
[the remark this answers is not in the document]
Comment (eric846) [reply source/adaptive_load/step_controller_impl.cc:108]: Definitely, filed #472 for that.

File: source/adaptive_load/step_controller_impl.cc [hunk 5]
```diff
@@ -0,0 +1,161 @@
...
+    nighthawk::client::CommandLineOptions command_line_options_template)
+    : command_line_options_template_{std::move(command_line_options_template)}, is_exponential_phase_{true},
+      exponential_factor_{config.exponential_factor() > 0.0 ? config.exponential_factor() : 2.0},
+      previous_load_value_{std::numeric_limits<double>::signaling_NaN()},
+      current_load_value_{config.initial_value()},
+      bottom_load_value_{std::numeric_limits<double>::signaling_NaN()},
+      top_load_value_{std::numeric_limits<double>::signaling_NaN()} {
+  doom_reason_ = "";
+  if (config.has_input_variable_setter()) {
+    // Load failures should have been caught during input validation.
```
[the remark this answers is not in the document]
Comment (eric846) [reply source/adaptive_load/step_controller_impl.cc:92]: Added a RELEASE_ASSERT here.

We are definitely in need of some StatusOr helpers. Filed #474. There might be lot of template magic that we have to test separately.

File: source/adaptive_load/step_controller_impl.cc [hunk 6]
```diff
@@ -0,0 +1,161 @@
...
+
+} // namespace
+
+Envoy::ProtobufTypes::MessagePtr
+ExponentialSearchStepControllerConfigFactory::createEmptyConfigProto() {
+  return std::make_unique<ExponentialSearchStepControllerConfig>();
+}
+
+std::string ExponentialSearchStepControllerConfigFactory::name() const {
+  return "nighthawk.exponential-search";
```
[the remark this answers is not in the document]
Comment (eric846) [reply source/adaptive_load/step_controller_impl.cc:48]: Done. Filed #475 to do in other files.

File: test/adaptive_load/fake_plugins/fake_input_variable_setter/fake_input_variable_setter.cc [hunk 7]
```diff
@@ -0,0 +1,70 @@
...
+    : adjustment_factor_{config.adjustment_factor() > 0 ? config.adjustment_factor() : 1} {}
+
+absl::Status FakeInputVariableSetter::SetInputVariable(
+    nighthawk::client::CommandLineOptions& command_line_options, double input_value) {
+  command_line_options.mutable_connections()->set_value(static_cast<unsigned int>(input_value * adjustment_factor_));
+  return absl::OkStatus();
+}
+
+std::string FakeInputVariableSetterConfigFactory::name() const {
+  return "nighthawk.fake-input-variable-setter";
```
[the remark this answers is not in the document]
Comment (eric846) [reply test/adaptive_load/fake_plugins/fake_input_variable_setter/fake_input_variable_setter.cc]: Done

File: test/adaptive_load/step_controller_test.cc (hunk 1 above)
Comment (eric846) [reply test/adaptive_load/step_controller_test.cc]: Done, actually realized that the score could be the arg.

File: source/adaptive_load/step_controller_impl.cc (hunk 3 above)
[the remark this answers is not in the document]
Comment (oschaaf) [reply source/adaptive_load/step_controller_impl.cc]: That updated error message looks really helpful now, great.

File: test/adaptive_load/fake_plugins/fake_input_variable_setter/fake_input_variable_setter.cc [hunk 8]
```diff
@@ -0,0 +1,70 @@
...
+  return absl::Status(static_cast<absl::StatusCode>(status_proto.code()), status_proto.message());
+}
+
+} // namespace
+
+FakeInputVariableSetter::FakeInputVariableSetter(
+    const nighthawk::adaptive_load::FakeInputVariableSetterConfig& config)
+    : adjustment_factor_{config.adjustment_factor() > 0 ? config.adjustment_factor() : 1} {}
+
+absl::Status FakeInputVariableSetter::SetInputVariable(
```
Comment (oschaaf) [inline test/adaptive_load/fake_plugins/fake_input_variable_setter/fake_input_variable_setter.cc:17]: const double?

File: test/adaptive_load/fake_plugins/fake_input_variable_setter/fake_input_variable_setter.cc [hunk 9]
```diff
@@ -0,0 +1,70 @@
...
+    nighthawk::client::CommandLineOptions& command_line_options, double input_value) {
+  command_line_options.mutable_connections()->set_value(static_cast<unsigned int>(input_value * adjustment_factor_));
```
Comment (oschaaf) [inline test/adaptive_load/fake_plugins/fake_input_variable_setter/fake_input_variable_setter.cc:19]: I wonder -- what happens when input value is negative? Do we need special handling? Is this covered in a test? If it is not possible to get negative input, maybe assert here?

File: test/adaptive_load/fake_plugins/fake_input_variable_setter/fake_input_variable_setter.cc [hunk 10]
```diff
@@ -0,0 +1,70 @@
+#include "test/adaptive_load/fake_plugins/fake_input_variable_setter/fake_input_variable_setter.h"
+
+namespace Nighthawk {
+
+namespace {
+
+absl::Status StatusFromProtoRpcStatus(const google::rpc::Status& status_proto) {
+  return absl::Status(static_cast<absl::StatusCode>(status_proto.code()), status_proto.message());
```
Comment (oschaaf) [inline test/adaptive_load/fake_plugins/fake_input_variable_setter/fake_input_variable_setter.cc:8]: Could this throw? If so, who should handle that?

File: test/adaptive_load/fake_plugins/fake_input_variable_setter/fake_input_variable_setter.cc [hunk 11]
```diff
@@ -0,0 +1,70 @@
...
+  } catch (const Envoy::EnvoyException& e) {
+    return absl::InvalidArgumentError(
+        absl::StrCat("Failed to parse FakeInputVariableSetterConfig proto: ", e.what()));
+  }
+}
+
+REGISTER_FACTORY(FakeInputVariableSetterConfigFactory, InputVariableSetterConfigFactory);
+
+envoy::config::core::v3::TypedExtensionConfig
+MakeFakeInputVariableSetterConfig(int adjustment_factor) {
```
Comment (oschaaf) [inline test/adaptive_load/fake_plugins/fake_input_variable_setter/fake_input_variable_setter.cc:59]: const the adjustment factor?

File: test/adaptive_load/fake_plugins/fake_input_variable_setter/fake_input_variable_setter.proto [hunk 12]
```diff
@@ -0,0 +1,15 @@
...
+package nighthawk.adaptive_load;
+
+import "google/rpc/status.proto";
+
+// Configuration for FakeInputVariableSetter (plugin name: "nighthawk.fake_input_variable_setter")
+// that sets the |connections| field in CommandLineOptions.
+message FakeInputVariableSetterConfig {
+  // A factor that multiplies the input value before applying it, in order to test the propagation
+  // of both the input and the config. Optional, default 1.
+  int32 adjustment_factor = 1;
```
Comment (oschaaf) [inline test/adaptive_load/fake_plugins/fake_input_variable_setter/fake_input_variable_setter.proto:12]: Consistency nit: I think we've been mostly using int32_t in other code

**Review by oschaaf — CHANGES_REQUESTED** [review]
Last feedback from me; additionally from the line-level comments I think there might be some repeated hard coded numbers in the tests that could be defined as named constants. Other then this LGTM

File: test/adaptive_load/fake_plugins/fake_input_variable_setter/fake_input_variable_setter.cc (hunk 9 above)
Comment (oschaaf) [reply test/adaptive_load/fake_plugins/fake_input_variable_setter/fake_input_variable_setter.cc:19]: Sorry in retrospect this commebr seems overly strict as I just realize this is test code; but maybe this is worth broader consideration?

File: test/adaptive_load/fake_plugins/fake_input_variable_setter/fake_input_variable_setter.proto (hunk 12 above)
Comment (eric846) [reply test/adaptive_load/fake_plugins/fake_input_variable_setter/fake_input_variable_setter.proto:12]: In proto we only seem to have int32. Also changing this to uint32 since we don't need negative values. Previously the artificial validation errors were generated by setting a negative adjustment_factor, but now we set the artificial error explicitly.

File: test/adaptive_load/fake_plugins/fake_input_variable_setter/fake_input_variable_setter.cc (hunk 9 above)
Comment (eric846) [reply test/adaptive_load/fake_plugins/fake_input_variable_setter/fake_input_variable_setter.cc:19]: Just made it uint32 since we don't need negative values.

File: test/adaptive_load/fake_plugins/fake_input_variable_setter/fake_input_variable_setter.cc (hunk 10 above)
Comment (eric846) [reply test/adaptive_load/fake_plugins/fake_input_variable_setter/fake_input_variable_setter.cc:8]: This was interesting... It turns out that casting an integer to an enum outside the defined range doesn't throw but produces undefined behavior. And C++ has no maintainable way to check for the largest value in this enum -- absl::StatusCode has a sentinel value that they say not to use, and it won't work anyway because of the gap from 16 to 20 [0].

Everyone who uses this field is expected to use standard Google codes, maybe through constants like grpc::DATA_LOSS.

I'm adding a warning about the enum issue to the proto comment. Do you think that's enough?

A safer but stranger alternative would be to define this field as not supporting anything above kAuthenticated=16, not tracking new absl::StatusCode values as they are added.

[0] https://github.com/abseil/abseil-cpp/blob/fbf0fdab62ac6fcc72b935d505f6837e0884fdc3/absl/status/status.h#L159

File: test/adaptive_load/fake_plugins/fake_input_variable_setter/fake_input_variable_setter.cc (hunk 8 above)
Comment (eric846) [reply test/adaptive_load/fake_plugins/fake_input_variable_setter/fake_input_variable_setter.cc:17]: Since this is passed by value we probably don't need const.

File: test/adaptive_load/fake_plugins/fake_input_variable_setter/fake_input_variable_setter.cc (hunk 11 above)
Comment (eric846) [reply test/adaptive_load/fake_plugins/fake_input_variable_setter/fake_input_variable_setter.cc]: Since it's passed by value I don't think we need const.

File: source/adaptive_load/input_variable_setter_impl.cc [hunk 13]
```diff
@@ -37,6 +37,11 @@ InputVariableSetterPtr RequestsPerSecondInputVariableSetterConfigFactory::create
   return std::make_unique<RequestsPerSecondInputVariableSetter>(config);
 }
 
+absl::Status RequestsPerSecondInputVariableSetterConfigFactory::ValidateConfig(
+    const Envoy::Protobuf::Message&) const {
+  return absl::OkStatus();
```
[the remark this answers is not in the document]
Comment (eric846) [reply source/adaptive_load/input_variable_setter_impl.cc:42]: See new comment in config_validator.h that explains the current contract here.  The full story  is a bit complicated and not ideal because of the mixture of these weak types, exceptions, and Status.

The short answer is that I intentionally don't want the ValidateConfig author to be responsible for checking the type because it's already checked elsewhere.

Also adding a comment to issue #468 about potential template gymnastics to make factories more explicit about types.

What do you think?

File: source/adaptive_load/step_controller_impl.cc [hunk 14]
```diff
@@ -0,0 +1,161 @@
...
+#include "api/adaptive_load/step_controller_impl.pb.h"
+
+#include "adaptive_load/input_variable_setter_impl.h"
+#include "adaptive_load/plugin_loader.h"
+
+namespace Nighthawk {
+
+namespace {
+
+using nighthawk::adaptive_load::BenchmarkResult;
```
[the remark this answers is not in the document]
Comment (eric846) [reply source/adaptive_load/step_controller_impl.cc:19]: Done.

File: source/adaptive_load/step_controller_impl.cc [hunk 15]
```diff
@@ -0,0 +1,161 @@
...
+}
+
+absl::Status ExponentialSearchStepControllerConfigFactory::ValidateConfig(
+    const Envoy::Protobuf::Message& message) const {
+  const auto& any = dynamic_cast<const Envoy::ProtobufWkt::Any&>(message);
+  ExponentialSearchStepControllerConfig config;
+  Envoy::MessageUtil::unpackTo(any, config);
+  if (config.has_input_variable_setter()) {
+    absl::StatusOr<InputVariableSetterPtr> input_variable_setter_or = LoadInputVariableSetterPlugin(config.input_variable_setter());
+    if (!input_variable_setter_or.ok()) {
```
[the remark this answers is not in the document]
Comment (eric846) [reply source/adaptive_load/step_controller_impl.cc:59]: Hah! That got a lot shorter.

File: source/adaptive_load/step_controller_impl.cc [hunk 16]
```diff
@@ -0,0 +1,161 @@
...
+      return input_variable_setter_or.status();
+    }
+  }
+  return absl::OkStatus();
+}
+
+StepControllerPtr ExponentialSearchStepControllerConfigFactory::createStepController(
+    const Envoy::Protobuf::Message& message,
+    const nighthawk::client::CommandLineOptions& command_line_options_template) {
+  const auto& any = dynamic_cast<const Envoy::ProtobufWkt::Any&>(message);
```
[the remark this answers is not in the document]
Comment (eric846) [reply source/adaptive_load/step_controller_impl.cc:65]: Yes, definitely, added to #468.

File: source/adaptive_load/step_controller_impl.h [hunk 17]
```diff
@@ -0,0 +1,57 @@
...
+  absl::StatusOr<nighthawk::client::CommandLineOptions>
+  GetCurrentCommandLineOptions() const override;
+  bool IsConverged() const override;
+  bool IsDoomed(std::string& doom_reason) const override;
+  void UpdateAndRecompute(const nighthawk::adaptive_load::BenchmarkResult& result) override;
+
+private:
+  const nighthawk::client::CommandLineOptions command_line_options_template_;
+  InputVariableSetterPtr input_variable_setter_;
+  bool is_exponential_phase_;
```
[the remark this answers is not in the document]
Comment (eric846) [reply source/adaptive_load/step_controller_impl.h:31]: This is actually the standard name, if it can be called that. I was surprised to find out about the binary search that is part of it: https://en.wikipedia.org/wiki/Exponential_search

File: source/adaptive_load/step_controller_impl.h [hunk 18]
```diff
@@ -0,0 +1,57 @@
...
+      nighthawk::client::CommandLineOptions command_line_options_template);
+  absl::StatusOr<nighthawk::client::CommandLineOptions>
+  GetCurrentCommandLineOptions() const override;
+  bool IsConverged() const override;
+  bool IsDoomed(std::string& doom_reason) const override;
+  void UpdateAndRecompute(const nighthawk::adaptive_load::BenchmarkResult& result) override;
+
+private:
+  const nighthawk::client::CommandLineOptions command_line_options_template_;
+  InputVariableSetterPtr input_variable_setter_;
```
[the remark this answers is not in the document]
Comment (eric846) [reply source/adaptive_load/step_controller_impl.h:34]: Added comments.

File: test/adaptive_load/fake_plugins/fake_input_variable_setter/fake_input_variable_setter_test.cc [hunk 19]
```diff
@@ -0,0 +1,129 @@
...
+  Envoy::ProtobufTypes::MessagePtr message = config_factory.createEmptyConfigProto();
+  FakeInputVariableSetterConfig expected_config;
+  EXPECT_EQ(message->DebugString(), expected_config.DebugString());
+}
+
+TEST(FakeInputVariableSetterConfigFactory, FactoryRegistersUnderCorrectName) {
+  FakeInputVariableSetterConfig config;
+  Envoy::ProtobufWkt::Any config_any;
+  config_any.PackFrom(config);
+  CommandLineOptions options;
```
[the remark this answers is not in the document]
Comment (eric846) [reply test/adaptive_load/fake_plugins/fake_input_variable_setter/fake_input_variable_setter_test.cc:33]: Removed

File: test/adaptive_load/step_controller_test.cc [hunk 20]
```diff
@@ -0,0 +1,194 @@
...
+  Envoy::ProtobufTypes::MessagePtr message = config_factory.createEmptyConfigProto();
+  nighthawk::adaptive_load::ExponentialSearchStepControllerConfig expected_config;
+  EXPECT_EQ(message->DebugString(), expected_config.DebugString());
+}
+
+TEST(ExponentialSearchStepControllerConfigFactory, CreatesCorrectFactoryName) {
+  nighthawk::adaptive_load::ExponentialSearchStepControllerConfig config;
+  Envoy::ProtobufWkt::Any config_any;
+  config_any.PackFrom(config);
+  nighthawk::client::CommandLineOptions options;
```
[the remark this answers is not in the document]
Comment (eric846) [reply test/adaptive_load/step_controller_test.cc:51]: Removed

File: test/adaptive_load/step_controller_test.cc [hunk 21]
```diff
@@ -0,0 +1,194 @@
...
+  auto& config_factory =
+      Envoy::Config::Utility::getAndCheckFactoryByName<StepControllerConfigFactory>(
+          "nighthawk.exponential-search");
+  StepControllerPtr plugin = config_factory.createStepController(config_any, options);
+  EXPECT_NE(dynamic_cast<ExponentialSearchStepController*>(plugin.get()), nullptr);
+}
+
+TEST(ExponentialSearchStepController, UsesInitialRps) {
+  nighthawk::adaptive_load::ExponentialSearchStepControllerConfig config;
+  config.set_initial_value(100.0);
```
[the remark this answers is not in the document]
Comment (eric846) [reply test/adaptive_load/step_controller_test.cc:72]: InputVariableSetter is a one-size-fits-all interface, so it always uses double. RequestsPerSecondInputVariableSetter casts the value to int and sets the integer proto field, so 100.5 will become 100.

I think the exponential search controller will actually routinely compute suggest non-integer values. It should all be ok since every InputVariableSetter is explicitly written to be able to deal with a double. Some future InputVariableSetter could want a precise fractional value

File: test/adaptive_load/fake_plugins/fake_input_variable_setter/fake_input_variable_setter.proto (hunk 12 above)
[the remark this answers is not in the document]
Comment (eric846) [reply test/adaptive_load/fake_plugins/fake_input_variable_setter/fake_input_variable_setter.proto]: This is independent of initial_value, but it could really go either way. The only purpose of this field is to confirm the propagation of the information, and an integer is sufficient for that. It's entirely a question of what seems more natural. What do you think?

**Comment by eric846** [comment]
> Last feedback from me; additionally from the line-level comments I think there might be some repeated hard coded numbers in the tests that could be defined as named constants. Other then this LGTM

Thanks for the review, created named constants and eliminated some unnecessary numbers.

File: test/adaptive_load/fake_plugins/fake_input_variable_setter/fake_input_variable_setter.proto (hunk 12 above)
Comment (oschaaf) [reply test/adaptive_load/fake_plugins/fake_input_variable_setter/fake_input_variable_setter.proto]: Oh my bad, this is a proto and not c++. Sorry about the noise

File: test/adaptive_load/fake_plugins/fake_input_variable_setter/fake_input_variable_setter.cc (hunk 10 above)
Comment (oschaaf) [reply test/adaptive_load/fake_plugins/fake_input_variable_setter/fake_input_variable_setter.cc:8]: My 2p would be: If the input can be trusted I think the comment suffices, and we should probably mention that in a comment explicitly. Hopefully ubsan in CI will have our back here too. If the input can't be trusted, we may need to think about how we can ensure defined behavior.. maybe then the "strange" alternative is the better option.

**Review by oschaaf — APPROVED** [review]
There's a pending question regarding StatusFromProtoRpcStatus, but as this is testcode it is not critical and it is probably fine  to discuss later. If we anticipate that prod code from other plugins might need this helper too then this may become more important to address. 

Right now this LGTM

**Comment by oschaaf** [comment]
Oh one last ask -- the coverage dipped below the threshold, it's not a flake this time. Is that something we can reasonably address? Or should we lower that threshold?

File: source/adaptive_load/step_controller_impl.cc [hunk 22]
```diff
@@ -0,0 +1,167 @@
...
+  return std::make_unique<ExponentialSearchStepController>(config, command_line_options_template);
+}
+
+REGISTER_FACTORY(ExponentialSearchStepControllerConfigFactory, StepControllerConfigFactory);
+
+ExponentialSearchStepController::ExponentialSearchStepController(
+    const ExponentialSearchStepControllerConfig& config,
+    nighthawk::client::CommandLineOptions command_line_options_template)
+    : command_line_options_template_{std::move(command_line_options_template)},
+      is_exponential_phase_{true}, exponential_factor_{config.exponential_factor() > 0.0
```
[the remark this answers is not in the document]
Comment (eric846) [reply source/adaptive_load/step_controller_impl.cc]: Done

File: source/adaptive_load/step_controller_impl.cc [hunk 23]
```diff
@@ -126,42 +121,55 @@ void ExponentialSearchStepController::UpdateAndRecompute(const BenchmarkResult&
...
-      top_load_value_ = current_load_value_;
-
-      previous_load_value_ = current_load_value_;
-      current_load_value_ = (bottom_load_value_ + top_load_value_) / 2;
-    }
+/**
+ * Updates state variables based on the latest score. Exponentially increases the load in each step.
+ * Transitions to the binary search phase when the load has caused metrics to go outside thresholds.
+ */
+void ExponentialSearchStepController::IterateRangeFindingPhase(double score) {
```
[the remark this answers is not in the document]
Comment (eric846) [reply source/adaptive_load/step_controller_impl.cc:135]: Tried that, but it needs to update fields.

File: source/adaptive_load/step_controller_impl.cc [hunk 24]
```diff
@@ -126,42 +121,55 @@ void ExponentialSearchStepController::UpdateAndRecompute(const BenchmarkResult&
...
+
     previous_load_value_ = current_load_value_;
     current_load_value_ = (bottom_load_value_ + top_load_value_) / 2;
   }
 }
 
+/**
+ * Updates state variables based on the latest score. Performs one step of a binary search.
+ */
+void ExponentialSearchStepController::IterateBinarySearchPhase(double score) {
```
[the remark this answers is not in the document]
Comment (eric846) [reply source/adaptive_load/step_controller_impl.cc:163]: Tried that, but it needs to update fields.
